mitodl/micromasters
# Zendesk prepopulate program on program page

Partial fix for #1262. This pull request builds on #1482 to also prepopulate the program field in the Zendesk ticket submission form. This pull request does several things:

* ~~Adds a `slug` field to the Program model, which is used to note which Zendesk field value this program is associated with~~
* Adds a `ProgramPageSerializer` class, which is used to serialize information about the program into the `SETTINGS` object in the template
* Refactors the callbacks in `zendesk_widget.js` to reflect the fact that Zendesk inserts _four_ iframes into the page, not just one
* Adds a `ticketSubmissionFormLoaded` callback so that, if we are on a program page, the program field dropdown will be set to whatever value matches the program on the page

Note that this pull request _only_ pre-populates the program field on a program page. No other page will cause the program field in the Zendesk form to pre-populate. A future pull-request can do that.

Also note that by including the faculty list and course list in the `ProgramPageSerializer`, we get some duplicated information in the `SETTINGS` object. I will clean up this duplication in a future pull request, as I did with  #1482 and #1615.

**How should this be manually tested?**

Go to the Programs model in the Django Admin (`/admin/courses/program/`). Create or edit a program so that the `slug` value is "scm". Next, make sure there is a ProgramPage for this program in Wagtail (`/cms`). If there isn't, go to the "Explorer" section in Wagtail, add a child page, and set the "Program" dropdown to the program you just edited in the Django admin. Visit the live page in your browser -- you'll need to publish the page in Wagtail before you can visit it. Once the page is loaded, click on the "Help" button at the bottom right of the page. Verify that the "MicroMasters Program" field in the ticket submission form is already set to "Supply Chain Management".

Next, go back to the Django Admin, and change the `slug` value to "dedp". Go back to the program page for this program, refresh the page, and click the "Help" button again. This time, the "MicroMasters Program" field in the ticket submission form should already be set to "Data, Economics, and Development Policy".

Finally, go back to the Django Admin, and change the `slug` value to any other string. Go back to the program page for this program, refresh the page, and click the "Help" button again. The "MicroMasters Program" field in the form should be set to the default value of "-".
